src/workerd/io/actor-cache.c++

```diff
@@ -573,61 +617,84 @@ kj::OneOf<ActorCache::GetResultList, kj::Promise<ActorCache::GetResultList>>
+  kj::Vector<kj::Promise<void>> promises;
+  for (size_t i = 0; i < getFlush.batches.size(); ++i) {
+    auto [p, f] = kj::newPromiseAndFulfiller<void>();
+    promises.add(kj::mv(p));
+
+    // We're going to do a getMultiple on this batch.
+    auto batch = getFlush.batches[i];
+    auto req = client.getMultipleRequest(capnp::MessageSize{
+      .wordCount = batch.wordCount,
+      .capCount = 0,
+    });
 
-  auto sendPromise = scheduleStorageRead(
-      [sizeHint,streamClient,&streamServerRef]
-      (rpc::ActorStorage::Operations::Client client) mutable -> kj::Promise<void> {
-    if (streamServerRef.nextExpectedKey == streamServerRef.keysToFetch.end()) {
-      // No more keys expected, must have finished listing on a previous try.
-      return kj::READY_NOW;
-    }
-    auto req = client.getMultipleRequest(sizeHint);
-    auto keysToFetch = kj::arrayPtr(streamServerRef.nextExpectedKey,
-                                    streamServerRef.keysToFetch.end());
-    auto list = req.initKeys(keysToFetch.size());
-    for (auto i: kj::indices(keysToFetch)) {
-      list.set(i, keysToFetch[i].asBytes());
+    kj::Vector<kj::Own<Entry>> batchEntries(batch.pairCount);
+    auto list = req.initKeys(batch.pairCount);
+    for (size_t j = 0; j < batch.pairCount; ++j) {
+      auto& entry = **entryIt;
+      ++entryIt;
+
+      list.set(j, entry.key.asBytes());
+      batchEntries.add(kj::atomicAddRef(entry));
     }
+
+    // The storage-proxy will stream results to us by calling `GetMultiStreamImpl::values()`,
+    // where we will update the `Entry`s in `batchEntries` with their values from storage (or set
+    // them ABSENT if they weren't found).
+    rpc::ActorStorage::ListStream::Client streamClient =
+        kj::heap<GetMultiStreamImpl>(*this, batchEntries.releaseAsArray(), kj::mv(f));
     req.setStream(streamClient);
-    return req.send().ignoreResult();
-  });
+    promises.add(req.send().ignoreResult().eagerlyEvaluate([](auto e) -> kj::Promise<void>{
+      return kj::mv(e);
+    }));
+  }
+  KJ_DASSERT(entryIt == getFlush.entries.end());
 
-  // Wait on the RPC only until stream.end() is called, then report the results. We prevent
-  // `stream` from being destroyed until we have a result so that if the RPC throws an exception,
-  // we don't accidentally report "PromiseFulfiller not fulfilled" instead of the exception.
-  auto promise = sendPromise
-      .then([&streamServerRef]() -> kj::Promise<ActorCache::GetResultList> {
-    if (streamServerRef.fulfiller->isWaiting()) {
-      return KJ_EXCEPTION(FAILED, "getMultiple() never called stream.end()");
-    } else {
-      // We'll be canceled momentarily...
-      return kj::NEVER_DONE;
-    }
+  // After this point, the only strong references to `Entry`s should be in the individual
+  // GetMultiStreamImpl objects.
+  getFlush.batches.clear();
+  getFlush.entries.clear();
+
+  // TODO(now): This doesn't seem to be firing...
```

Comment:
The `readCancellation` promise only gets resolved during the `ActorCache` destructor. That `cancelReadFlush->fulfill()` call there adds an event to the kj event loop to call any callbacks that were waiting on the `readCancellation` promise, but it doesn't immediately call them. The `ActorCache` finishes getting destructed (and whatever code path caused the destruction of the `ActorCache` would also continue running up until it yields the event loop) before any of those callbacks would run.

And, importantly, `syncGets()` is only called from within the `flushImpl()` code path, which is a chain of async work owned by the `lastFlush` promise on the `ActorCache` class. So when the `ActorCache` instance gets destructed, that `lastFlush` promise will also get canceled.

So it would surprise me greatly (and probably be a bad thing) if there was some code path in which this promise fired. Which means we may need a different approach to more synchronously do whatever needs doing here during destruction (assuming it can't be handled gracefully via destructors/RAII).

Reply:
https://github.com/cloudflare/workerd/pull/1916/commits/497ded47c634e0505299694de82f7cef780c8864

This seems to get the test passing locally, but some of our internal tests are seeing segfaults now (need to look into them more, this is weirdly tricky). Any idea why destroying the joint promise wouldn't also cause the `GetMultiStreamImpl` to be dropped? I had to explicitly clear out the `GetMultiStreamImpl::entries`, was hoping to just dtor the object but it seems like its lifetime is extended further, but I'm not sure what's keeping it alive.

Reply:
> Any idea why destroying the joint promise wouldn't also cause the GetMultiStreamImpl to be dropped?

If the call to `syncGets()` is still being awaited on somewhere, then the `clients` vector would still be getting held on the relevant coroutine stack. We could presumably fix that by either (A) moving the `clients` vector into an attachment on the joint promise, or (B) making sure that whatever promise that's keeping the `syncGets` call alive gets canceled.

And I'd guess what's keeping it alive is the `lastFlush` promise, which is held as a member variable on `ActorCache`.

But would you like me to try poking at this locally? It's always easier to figure cancellation / destruction order stuff out by testing stuff out than just by looking at the code and guessing.

Reply:
So, the test that would fail is `bazel test @workerd//src/workerd/io:actor-cache-test`, specifically `ActorCache list stream cancellation`. It seems like the joint promise does get cancelled (and the individual promises in the loop also get cancelled), but the `GetMultiStreamImpl`, which we move into a `rpc::ActorStorage::ListStream::Client`, doesn't seem to get destroyed.

I'm not sure if it's because the `ListStream::Client` is living long, or if the `req` object which takes a reference to the `ListStream::Client` is living long. I suspect we're somehow creating a strong reference to the capability (briefly looking at `req.setStream()`, I see we eventually call `Capability::Client::hook->addRef()` in `PointerHelpers<>::set()`), preventing it from being dtor'd even though the promise is done.

Reply:
It looks like this is fixed now? At least it's not failing for me when I pull this branch down and try to repro it.

Reply:
Yup! This test is fine, but one of our internal tests is failing because of a segfault that was introduced when fixing `ActorCache list stream cancellation`. I'll let you know which one some time this week, not able to connect to wifi on my laptop for some reason.

When we destroy `ActorCache`, we go through and destroy the promises in `syncGets()`, which results in us calling `GetMultiStreamImpl::clear()`. However, sometimes this object has already been destroyed by the time we call `clear()`. There's probably a reasonable fix, but this already feels a bit hacky, which is why I mentioned that it'd be better to find a more robust approach.

Reply:
https://github.com/cloudflare/workerd/pull/1916/commits/875e4aca6dfc652578cf2c02b04762b4071d658a fixes the segfault and I think is sufficiently commented on.